### ISSUE TITLE
[REVIEW]: 𝚗𝚛𝙲𝚊𝚜𝚌𝚊𝚍𝚎𝚂𝚒𝚖 -- A simulation tool for nuclear recoil cascades resulting from neutron capture

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3993]
**Submitting author:** @villaa (<a href="http://orcid.org/0000-0002-3893-7259">Anthony Villano</a>)
**Repository:** https://github.com/villano-lab/nrCascadeSim
**Branch with paper.md** (empty if default branch): 
**Version:** v1.4.0
**Editor:** @lucydot
**Reviewers:** @matthewfeickert, @altavir
**Archive:** 10.5281/zenodo.6056681

 

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/d69ced49c5c17fdbf637e0747d815deb/status.svg)](https://joss.theoj.org/papers/d69ced49c5c17fdbf637e0747d815deb)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/d69ced49c5c17fdbf637e0747d815deb"><img src="https://joss.theoj.org/papers/d69ced49c5c17fdbf637e0747d815deb/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/d69ced49c5c17fdbf637e0747d815deb/status.svg)](https://joss.theoj.org/papers/d69ced49c5c17fdbf637e0747d815deb)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/villano-lab/nrCascadeSim" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@matthewfeickert & @altavir, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @lucydot know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

## Review checklist for @matthewfeickert
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/villano-lab/nrCascadeSim">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/villano-lab/nrCascadeSim">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@villaa) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @altavir
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/villano-lab/nrCascadeSim">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/villano-lab/nrCascadeSim">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@villaa) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @matthewfeickert, @altavir it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
Wordcount for `paper.md` is 1201

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1515/iupac.70.0026 is OK
- 10.1103/PhysRevD.91.083509 is OK
- 10.1103/physrevc.43.1086 is OK
- 10.1103/PhysRevC.46.972 is OK
- 10.1016/S0168-9002(03)01368-8 is OK
- 10.1088/1748-0221/16/07/p07032 is OK
- 10.1103/PhysRevC.82.054616 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.18 s (591.1 files/s, 163342.4 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
JavaScript                       9           2404           2458           9198
C                               13            711            970           2536
SVG                              2              0             21           2248
CSS                              5            334             50           1293
C++                             11            254            252            963
reStructuredText                12            249            337            530
Python                           8            164            269            528
Markdown                        14            189              0            517
C/C++ Header                    10            154            308            484
HTML                             4            131              1            269
Jupyter Notebook                 2              0            508            133
TeX                              1             10              1            125
YAML                             6             16             38            109
make                             3             31             18             84
Bourne Shell                     4             13              6             65
DOS Batch                        1              8              1             26
-------------------------------------------------------------------------------
SUM:                           105           4668           5238          19108
-------------------------------------------------------------------------------


Statistical information for the repository 'e6613cefdb2052624e444cb9' was
gathered on 2021/12/14.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Anthony Villano                 11         21540            431           90.36
Kitty                           11          1089           1254            9.64

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Anthony Villano           19984           92.8          0.3               20.39
Kitty                       960           88.2          8.6               25.73
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03993/joss.03993/10.21105.joss.03993.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03993/joss.03993/10.21105.joss.03993.pdf) :page_facing_up: :point_left:

**Comment by matthewfeickert** [comment]
@lucydot @altavir As I already have done a few reviews of the project, and @villaa and the rest of the dev team are working on [Issues that I raised in the pre review](https://github.com/openjournals/joss-reviews/issues/3932#issuecomment-990735982) RE: release [`v1.2.3`](https://github.com/villano-lab/nrCascadeSim/releases/tag/v1.2.3) (c.f. https://github.com/openjournals/joss-reviews/issues/3932#issuecomment-993030180) I'm going to hold of on starting my review until they finish those and release [`v1.2.4`](https://github.com/villano-lab/nrCascadeSim/releases/tag/v1.2.4) and notify us here :+1:  (once that is done can we also get whedon to update the Version number in the Issue body?). I suspect that this will only take a few days as the team seemed quite eager to get the review started, which is great.

I'll of course take a look at the paper itself now, but my memory from [the previous review](https://github.com/openjournals/joss-reviews/issues/3174) was that it was basically fine, so I'm not expecting to have many, if any, comments there.

**Comment by matthewfeickert** [comment]
@whedon commands

**Comment by whedon** [comment]
Here are some things you can ask me to do:

```
# List Whedon's capabilities
@whedon commands

# List of editor GitHub usernames
@whedon list editors

# List of reviewers together with programming language preferences and domain expertise
@whedon list reviewers

EDITORIAL TASKS

# Compile the paper
@whedon generate pdf

# Compile the paper from alternative branch
@whedon generate pdf from branch custom-branch-name

# Ask Whedon to check the references for missing DOIs
@whedon check references

# Ask Whedon to check repository statistics for the submitted software
@whedon check repository

```

**Comment by villaa** [comment]
@matthewfeickert @altavir we have released `v1.2.4` which takes care of an obvious bug while running the `realizeCascades` executable. This should allow better testing for now. The team is also working on the other major thing revealed in the. pre-review, which is related to issue 27 in `nrCascadeSim`. When that is complete we expect to go to `v1.3.0` and I will notify here.

**Comment by lucydot** [comment]
That's excellent thanks @villaa and @matthewfeickert for the updates.

@altavir I can see you have also started working through the reviewer checklist which is great. I guess you've probably already seen the link to the previous review of this code, but just in case you missed it - it is here: https://github.com/openjournals/joss-reviews/issues/3174.

**Comment by altavir** [comment]
@lucydot thanks. I will get back to that as soon as I finish some urgent tasks.

**Comment by lucydot** [comment]
@altavir @villaa @matthewfeickert I am setting my out of office for the next two week Christmas / New Year period. I will check back in next week to see progress, but a heads up that I may not respond to messages within the typical 3 day period.

RE: updating the software version number for review. I can manually update the version number in the review thread at the top, but first will double check with other members of the editorial team that this won't have have any unwanted side effects. The latest software release for review is `v1.2.4` with an update to `v.1.3.0` expected (as outlined by @villaa in [the post above](https://github.com/openjournals/joss-reviews/issues/3993#issuecomment-995414078)).

**Comment by villaa** [comment]
@altavir @matthewfeickert @lucydot @gerudo7  -- the version `v1.3.0` has now been released. This is the best version for your further reviews and comments. There are some issues left from the pre-review but they were both explicitly said to be not required for the review (adding code coverage and using doxygen to auto-generate docs). 

We have upgraded so that OSX 10.14, 10.15 are now working on Travis-CI. 

We will work on the remaining issues concurrently with the other issues from the review and update here for any changes. I invite you to please give feedback for your review and thank you for agreeing to participate.

**Comment by whedon** [comment]
:wave: @matthewfeickert, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @altavir, please update us on how your review is going (this is an automated reminder).

**Comment by villaa** [comment]
@lucydot @matthewfeickert @altavir --  one last update. We have released  [`v1.3.1` ](https://github.com/villano-lab/nrCascadeSim/releases/tag/v1.3.1)and all issues from the previous review and pre-review are  addressed. We have test coverage of around 80%. 

If you have not started the review this is the best version to use. If you started on `v1.3.0` the only differences are upgraded documentation and code test coverage.

**Comment by lucydot** [comment]
Thanks for the update @villaa and Happy New Year ✨ 

I'll update the version number now. Usually we update the version number at the end of the review process - in this case, I'll update now to avoid any confusion as to which version is under review.

**Comment by lucydot** [comment]
@whedon set version as v1.3.1

**Comment by whedon** [comment]
I'm sorry human, I don't understand that. You can see what commands I support by typing:

```
@whedon commands
```

**Comment by lucydot** [comment]
@whedon set v1.3.1 as version

**Comment by whedon** [comment]
OK. v1.3.1 is the version.

**Comment by altavir** [comment]
Happy new year to everyone. I've finally found some time to review the code and the paper and open [some issues](https://github.com/villano-lab/nrCascadeSim/issues/created_by/altavir). Not all of them are critical though. What I think is the most important is [this one](https://github.com/villano-lab/nrCascadeSim/issues/75). The lack of comparison of results with other tools as well as regression tests could be a real problem in the future.

I also failed to build the project on WSL even with CERN ROOT installed ( I have no clue why it is needed anyway), so I can't comment on the program working (I will try later to run it again), but even from looking into the code there seems to some things that could be improved. I am not a C++ expert though.

**Comment by villaa** [comment]
@altavir Happy New Year! Thank you very much for your review,  I see the issues on our issue tracker and the team will address them and comment here when they're finished. 

Do you know what style of Linux WSL uses? We haven't tested on Windows but could try to match the system. 

@gerudo7 do you run WSL?

**Comment by gerudo7** [comment]
> @gerudo7 do you run WSL?
Yes! I've been doing all my work with nrCascadeSim in WSL. I know for certain that it works for me with root version 6.22/00 installed via conda, but I just tried it with the latest precompiled binary for version 6.24 and nrCascadeSim won't make. I'm guessing that the problem is with using the precompiled binaries in WSL, but I'm still testing that theory. (I will work on adding whatever I find to the docs as well.)

**Comment by altavir** [comment]
I am using Root 6.24.0 installed via nix-env. I think it is better to continue the discussion in your repository, You can open an issue and we can discuss it there.

**Comment by matthewfeickert** [comment]
Just popping in to say that I'm back from holidays now and so will try to get my review in before Wednesday. I see that the authors have done a lot of nice work, which is great to see, and I look forward to getting to review it. :slightly_smiling_face:

**Comment by villaa** [comment]
> Just popping in to say that I'm back from holidays now and so will try to get my review in before Wednesday. I see that the authors have done a lot of nice work, which is great to see, and I look forward to getting to review it. 🙂

@matthewfeickert Do you have an update?

**Comment by matthewfeickert** [comment]
> Do you have an update?

No. Maybe tonight. High probability before Monday (2022-01-17).

**Comment by matthewfeickert** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03993/joss.03993/10.21105.joss.03993.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03993/joss.03993/10.21105.joss.03993.pdf) :page_facing_up: :point_left:

**Comment by altavir** [comment]
The primary critical point that should be fixed in my opinion is https://github.com/villano-lab/nrCascadeSim/issues/75. The part with comparison. Data tests would be welcome for future development, but not so critical.

**Comment by villaa** [comment]
> The primary critical point that should be fixed in my opinion is [villano-lab/nrCascadeSim#75](https://github.com/villano-lab/nrCascadeSim/issues/75). The part with comparison. Data tests would be welcome for future development, but not so critical.

@altavir thanks, we are working on using Geant4 model for our comparison. It is taking a bit of time b/c we want to get the most recent version and I haven't run the module in a while. Incidentally, [here](https://github.com/villano-lab/G4PhysicsTesting) is the code we're using for that.

**Comment by matthewfeickert** [comment]
Thanks for your submission and software @villaa. The following is my review of it, which goes into detail (and so isn't just noted in the first post of this Issue) only to try to help make requested revisions more actionable and clear. As always, the review is following the [JOSS paper review criteria](https://joss.readthedocs.io/en/latest/review_criteria.html). Also, please feel free to ask for clarification on anything as I'd like to make sure that this information is helpful and actionable to you. :+1:

For consistency and reproducibility of all my comments, I will be using the public Docker image `atlasamglab/stats-base:root6.24.06` (where ROOT is built from source and installed from the Git tag `6.24.06`) to conduct my review which has the following software specs

```console
$ docker run --rm -ti -v $PWD:$PWD -w $PWD atlasamglab/stats-base:root6.24.06
# gcc --version | head -n 1
gcc (Debian 10.2.1-6) 10.2.1 20210110
# root --version
ROOT Version: 6.24/06
Built for linuxx8664gcc on Jan 12 2022, 09:39:00
From tags/v6-24-06@v6-24-06
# root-config --cflags
-pthread -std=c++17 -m64 -I/usr/local/venv/include
```

## Review checklist

This review is for [tag `v1.3.1`](https://github.com/villano-lab/nrCascadeSim/releases/tag/v1.3.1) of the project.

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/villano-lab/nrCascadeSim">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?

   - The repo has an OSI approved license (MIT), but vendors the Msersenne Twister prime algorithm without any LICENSE, which is a problem. All vendored software must be properly licensed. It is also unclear why this is even needed as the [Mersenne Twister 19937 generator is in the C++ Standard Library](https://www.cplusplus.com/reference/random/mt19937/).

- [x] **Contribution and authorship:** Has the submitting author (@villaa) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

   - Yes. Out of the number of commits in this project related to the software itself the submitting author @villaa has made the majority of them, and team member @gerudo7 has made nearly all the rest.

   ```
   $ git log --oneline ":(exclude)paper" | wc -l
   507
   $ git log --author="Anthony Villano" --oneline ":(exclude)paper" | wc -l
   296
   $ git log --author="Kitty" --oneline ":(exclude)paper" | wc -l
   195
   ```

- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

JOSS asks for a minimum of 3 months of work equivalent effort for the software to be considered. The project definitely fulfills this requirement.

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
   - Yes. :+1:

<details>
<summary>Install example:</summary>

```console
# git clone --depth 1 https://github.com/villano-lab/nrCascadeSim.git --branch v1.3.1 --single-branch
# cd nrCascadeSim/
# make
g++ -fPIC -c -D__GIT_VERSION=\"v1.3.1\" -IMersenne -Iinc src//isotope_info.c  `root-config --cflags` -L. -L/home/data/nrCascadeSim/bin/lib -fprofile-arcs -ftest-coverage
mv isotope_info.o /home/data/nrCascadeSim/bin/lib/
g++ -fPIC -c -D__GIT_VERSION=\"v1.3.1\" -IMersenne -Iinc src//weisskopf.c  `root-config --cflags` -L. -L/home/data/nrCascadeSim/bin/lib -fprofile-arcs -ftest-coverage
mv weisskopf.o /home/data/nrCascadeSim/bin/lib/
g++ -fPIC -c -D__GIT_VERSION=\"v1.3.1\" -IMersenne -Iinc src//lindhard.c  `root-config --cflags` -L. -L/home/data/nrCascadeSim/bin/lib -fprofile-arcs -ftest-coverage
mv lindhard.o /home/data/nrCascadeSim/bin/lib/
g++ -fPIC -c -D__GIT_VERSION=\"v1.3.1\" -IMersenne -Iinc src//cascadeProd.c  `root-config --cflags` -L. -L/home/data/nrCascadeSim/bin/lib -fprofile-arcs -ftest-coverage
mv cascadeProd.o /home/data/nrCascadeSim/bin/lib/
g++ -fPIC -c -D__GIT_VERSION=\"v1.3.1\" -IMersenne -Iinc src//edepmath.c `root-config --cflags` -L. -L/home/data/nrCascadeSim/bin/lib -fprofile-arcs -ftest-coverage
mv edepmath.o /home/data/nrCascadeSim/bin/lib/
g++ -fPIC -c -D__GIT_VERSION=\"v1.3.1\" -IMersenne -Iinc src//rootUtil.c `root-config --cflags`  -L. -L/home/data/nrCascadeSim/bin/lib -fprofile-arcs -ftest-coverage
mv rootUtil.o /home/data/nrCascadeSim/bin/lib/
g++ -fPIC -shared /home/data/nrCascadeSim/bin/lib/lindhard.o /home/data/nrCascadeSim/bin/lib/weisskopf.o /home/data/nrCascadeSim/bin/lib/isotope_info.o /home/data/nrCascadeSim/bin/lib/cascadeProd.o /home/data/nrCascadeSim/bin/lib/edepmath.o /home/data/nrCascadeSim/bin/lib/rootUtil.o `root-config --cflags` -L`root-config --libdir` -lCore -lRIO -lTree  -o /home/data/nrCascadeSim/bin/lib/libncap.so -fprofile-arcs -ftest-coverage
g++ -fPIC -Wl,-rpath,`root-config --libdir`,-rpath /home/data/nrCascadeSim/bin/lib -D__GIT_VERSION=\"v1.3.1\" -IMersenne -Iinc -L. -L/home/data/nrCascadeSim/bin/lib /home/data/nrCascadeSim/bin/realizeCascades.cpp `root-config --cflags` -L`root-config --libdir` -lCore -lRIO -lTree -lncap -o /home/data/nrCascadeSim/bin/realizeCascades -coverage -fprofile-arcs -ftest-coverage
g++ -fPIC -Wl,-rpath /home/data/nrCascadeSim/bin/lib -D__GIT_VERSION=\"v1.3.1\" -IMersenne -Iinc -L. -L/home/data/nrCascadeSim/bin/lib /home/data/nrCascadeSim/bin/regexPlayground.cpp -o /home/data/nrCascadeSim/bin/regexPlayground -coverage -fprofile-arcs -ftest-coverage
root@4b4531b84c8f:~/data/nrCascadeSim# tree bin  # I install tree for example purposes
bin
├── lib
│   ├── cascadeProd.o
│   ├── edepmath.o
│   ├── isotope_info.o
│   ├── libncap.so
│   ├── lindhard.o
│   ├── rootUtil.o
│   └── weisskopf.o
├── realizeCascades
├── realizeCascades.cpp
├── regexPlayground
└── regexPlayground.cpp

1 directory, 11 files
root@4b4531b84c8f:~/data/nrCascadeSim# command -v realizeCascades
root@4b4531b84c8f:~/data/nrCascadeSim# make install
cp /home/data/nrCascadeSim/bin/realizeCascades /usr/local/bin/
cp /home/data/nrCascadeSim/bin/regexPlayground /usr/local/bin/
root@4b4531b84c8f:~/data/nrCascadeSim# command -v realizeCascades
/usr/local/bin/realizeCascades

```

</details>


The [`install` Make target](https://github.com/villano-lab/nrCascadeSim/blob/182e8a56406a8ae4e5cfc8df489d9779d25d93da/Makefile#L80-L82) hardcodes the install target directory to `/usr/local/bin`. Install locations can have defaults, but should allow for user input &mdash; there are many situations on remotes (example: CERN's LXPLUS) where users do not have access to `/usr/local/` and the users would need to manually edit the `Makefile`.

Also, the `clean` Make target should for the same reasons _not_ [try to `rm -f` anything under `/usr/local`](https://github.com/villano-lab/nrCascadeSim/blob/182e8a56406a8ae4e5cfc8df489d9779d25d93da/Makefile#L95-L96). This is also probably never what a user wants to have happen when `make clean` is run &mdash; `clean` targets are not uninstallers. Similarly, advocating directly for the use of `sudo` for installation of local software should be avoided. The `.gcno` generated during the build should also be cleaned in the `clean` target (also check for `.gcda` files to be safe).

A better alternative default is to assume the user is NOT root, and to attempt to install in the user's `~/bin` where they will be assured write access.

- [x] **Functionality:** Have the functional claims of the software been confirmed?

Yes. The software produces reasonable results as far as I can tell and is able to pass the test suite (which includes the claims for support of Neon, Argon, Silicon, and Germanium cascades).

- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

No performance claims are made.

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?

Yes. The docs include [the statement](https://github.com/villano-lab/nrCascadeSim/blob/182e8a56406a8ae4e5cfc8df489d9779d25d93da/docs/source/index.rst?plain=1#L34-L37)

> The purpose of this code is to simulate energy deposits due to cascading of energy levels following neutron capture. This code was written for use in nuclear recoil calibration for dark matter detectors, but may be useful in other particle physics applications as well, including coherent elastic neutrino nucleus scattering (CEνNS).

The [docs on ReadTheDocs](https://nrcascadesim.readthedocs.io/) are not versioned beyond `latest` and `stable`. It would be great to see releases have versioned docs on ReadTheDocs too.

- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.

   - Yes. The software only depends on the `ROOT` C++ library and the `gcc` compiler. It is not possible to install `ROOT` with a OS package manager (and probably never will be given the size of the software) but the team nicely mentions that there are pre-package binaries from the CERN ROOT team as well as a conda-forge distribution (Conda is a package/environment manager, so this counts in my mind).

- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).

Yes. I have tested the examples and found that they work. They also pertain to dark matter detection physics (the area of research expertise of the authors).

<details>
<summary>The main example from the top level of the repository:</summary>

```console
# time realizeCascades --verbose -n 100000 -o output.root levelfiles/Si28_ngam_all_cascades_rfmt_sorted.txt
levelfiles/Si28_ngam_all_cascades_rfmt_sorted.txt
Seed used: 498136696
MT Result: 0.424727

real	0m1.495s
user	0m1.345s
sys	0m0.098s
# root -l output.root
root [0]
Attaching file output.root as _file0...
(TFile *) 0x55d12fd98a40
root [1] _file0->ls()
TFile**		output.root
 TFile*		output.root
  KEY: TTree	cascade;1	cascade
root [2] cascade->Print()
******************************************************************************
*Tree    :cascade   : cascade                                                *
*Entries :    99968 : Total =        19053073 bytes  File  Size =    7391802 *
*        :          : Tree compression factor =   2.58                       *
******************************************************************************
*Br    0 :n         : n/L                                                    *
*Entries :    99968 : Total  Size=     802407 bytes  File Size  =       7376 *
*Baskets :       26 : Basket Size=      32000 bytes  Compression= 108.68     *
*............................................................................*
*Br    1 :cid       : cid/L                                                  *
*Entries :    99968 : Total  Size=     802467 bytes  File Size  =       6890 *
*Baskets :       26 : Basket Size=      32000 bytes  Compression= 116.35     *
*............................................................................*
*Br    2 :Elev      : Elev[n]/D                                              *
*Entries :    99968 : Total  Size=    2181158 bytes  File Size  =     142555 *
*Baskets :       81 : Basket Size=      32000 bytes  Compression=  15.29     *
*............................................................................*
*Br    3 :taus      : taus[n]/D                                              *
*Entries :    99968 : Total  Size=    2181158 bytes  File Size  =     142237 *
*Baskets :       81 : Basket Size=      32000 bytes  Compression=  15.32     *
*............................................................................*
*Br    4 :E         : E[n]/D                                                 *
*Entries :    99968 : Total  Size=    2180903 bytes  File Size  =    1141056 *
*Baskets :       81 : Basket Size=      32000 bytes  Compression=   1.91     *
*............................................................................*
*Br    5 :delE      : delE[n]/D                                              *
*Entries :    99968 : Total  Size=    2181158 bytes  File Size  =    1751576 *
*Baskets :       81 : Basket Size=      32000 bytes  Compression=   1.24     *
*............................................................................*
*Br    6 :I         : I[n]/D                                                 *
*Entries :    99968 : Total  Size=    2180903 bytes  File Size  =     421026 *
*Baskets :       81 : Basket Size=      32000 bytes  Compression=   5.18     *
*............................................................................*
*Br    7 :Ei        : Ei[n]/D                                                *
*Entries :    99968 : Total  Size=    2180988 bytes  File Size  =    1749733 *
*Baskets :       81 : Basket Size=      32000 bytes  Compression=   1.25     *
*............................................................................*
*Br    8 :time      : time[n]/D                                              *
*Entries :    99968 : Total  Size=    2181158 bytes  File Size  =    1880987 *
*Baskets :       81 : Basket Size=      32000 bytes  Compression=   1.16     *
*............................................................................*
*Br    9 :Eg        : Eg[n]/D                                                *
*Entries :    99968 : Total  Size=    2180988 bytes  File Size  =     142468 *
*Baskets :       81 : Basket Size=      32000 bytes  Compression=  15.29     *
*............................................................................*

```
</details>


The individual examples that are in the [`tests/bin` directory](https://github.com/villano-lab/nrCascadeSim/tree/182e8a56406a8ae4e5cfc8df489d9779d25d93da/bin) build and run, and from the documentation seem to be doing what they each describe.

The examples in [`example-usecase/Yields_and_Resolutions.ipynb`](https://github.com/villano-lab/nrCascadeSim/blob/182e8a56406a8ae4e5cfc8df489d9779d25d93da/example-usecase/Yields_and_Resolutions.ipynb) are nice and it would be great to see these get added into the docs using some of the Jupyter rendering Sphinx extensions.

- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?

Yes, the docs exist and while there are some Sphinx errors in terms of syntax, these are easy to fix.

There is an outline of the CLI API for the `realizeCascades` executable [that is hardcoded](https://github.com/villano-lab/nrCascadeSim/blob/182e8a56406a8ae4e5cfc8df489d9779d25d93da/docs/source/02_Executables.rst) and the [`nrCascadeSim` library docs are generated with Doxygen](https://github.com/villano-lab/nrCascadeSim/blob/182e8a56406a8ae4e5cfc8df489d9779d25d93da/docs/source/03_Doxy_Libraries.rst).

The `realizeCascades` text input file format and ROOT binary output file format are also documented (I might suggest promoting I/O to its own docs section).

The use of Doxygen for the library is nice (kudos so the authors for taking this pre-review comment under suggestion) and the only advice I have about hardcoding of CLI API is to make sure that all examples in the docs for that section are rigorously tested in CI.

- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?

Yes. The author have a CI system using Travis CI that tests the software against ROOT `v6.20.00` and `v6.24.06` across `c++11`, `c++14`, and `c++17` on various Ubuntu Linux distributions and a subset of those configurations on various MacOS releases.

The code coverage at the time of the review is 82%. While low, this is acceptable for the review (though I'd encourage the authors to keep up their nice work of improving the coverage).

As already mentioned in https://github.com/villano-lab/nrCascadeSim/issues/75, the code is lacking regressions tests. It would be good if there could be at least one in the test suite and if the authors could plan to add more in the future.

While changing CI providers is not something that is relevant for this review, it should be noted that Travis CI now (2020 onwards) [offers a open source projects a one time non-renewable 10,000 build credit](https://blog.travis-ci.com/2020-11-02-travis-ci-new-billing).
While it does not impact this review, the authors will need to consider how they plan to continue testing their software once this on time limit has been reached.

- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

Yes. There is a [`CONTRIBUTING.md`](https://github.com/villano-lab/nrCascadeSim/blob/182e8a56406a8ae4e5cfc8df489d9779d25d93da/CONTRIBUTING.md) and [`CODE_OF_CONDUCT.md`](https://github.com/villano-lab/nrCascadeSim/blob/182e8a56406a8ae4e5cfc8df489d9779d25d93da/CODE_OF_CONDUCT.md) and [the `README` mentions](https://github.com/villano-lab/nrCascadeSim/blob/182e8a56406a8ae4e5cfc8df489d9779d25d93da/README.md?plain=1#L51)

> For questions, support, bug reports, or other suggestions, please open an [issue](https://github.com/villano-lab/nrCascadeSim/issues).

Additionally, there is a [Contact & Support](https://nrcascadesim.readthedocs.io/en/stable/09_Contact.html) section of the docs and

```console
# realizeCascades --help | tail -n 3
Report bugs to: villaa-at-gmail-dot-com
realizeCascades (nrCascadeSim) home page: <https://github.com/villano-lab/nrCascadeSim>
General help nrCascadeSim Software: <https://nrcascadesim.readthedocs.io/en/latest/>
```

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?

Yes. The introduction does a good job of this.

- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?

Yes.

- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?

Yes. They compare `nrCascadeSim` to `Geant4` and `FIFRELIN` in a fair manner.

The authors don't mention this, but it is worth noting that [`Geant4` is a considerably larger project](https://github.com/Geant4/geant4) and given the large number of use cases it is meant to handle can be much more cumbersome to use. While `Geant4` is fully open source, it is not clear from searches if there are open source implementations of the `FIFRELIN` Monte Carlo simulator. If not, `nrCascadeSim` provides an advantage by being open source.

- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?

The paper is of good quality and of the proper scope following JOSS guidelines.


- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

Yes.

## Suggested revisions before JOSS publication

@lucydot, I think the paper and the project are in overall good shape. My recommendation is that once the following revisions are addressed (I'll open up GitHub Issues for them) the project be accepted for publication pending your approval. Though we also still need @altavir's written review as well.

### Code
* Change the `Makefile` to remove targeting `/usr/local` as a default. In many situations this is an unwritable location for users.
   - If https://github.com/villano-lab/nrCascadeSim/issues/64 is enacted this can be easily achieved with setting the [`CMAKE_INSTALL_PREFIX` option](https://cmake.org/cmake/help/latest/variable/CMAKE_INSTALL_PREFIX.html) during `cmake`.
* Add removal of the `.gcno` and `.gcda` files generated during the build to the `clean` Make target.
* Remove recommendations of use of `sudo` to install software. Anyone who needs to use `sudo` does not need to be told to use `sudo` and all software should also be able to be installed and used locally.
* Add at least one regression test to the test suite and plan to add more in the future (compliment to https://github.com/villano-lab/nrCascadeSim/issues/75).

#### Paper
* In the State of the Field section it would be beneficial to mention that `Genat4` is open source software and to clarify if `FIFRELIN` is open source or not.

## Suggested revisions outside of review

These suggestions should not be considered inside the scope of the review and editors should not view their completion as a dependency for approval.

* Fix all Sphinx errors and typos.
* Strive to increase code coverage. Code coverage is not a magic bullet, and even with 100% coverage there can still be integration errors. However, in my experience once you start getting code coverage going over 90% you can develop much more aggressively as you know that your test suites are doing the hard work for you. :rocket:
* Migrate off of Travis CI. Travis CI has decided it will no longer offer free testing to the open source community and the 10,000 credits it gives users on the free trial plan should be viewed as an off ramp to ensure that there is robust CI support. As mentioned in https://github.com/villano-lab/nrCascadeSim/issues/10, alternatives that are easy to switch to are GitHub Actions or Azure Pipelines.
* The [docs on ReadTheDocs](https://nrcascadesim.readthedocs.io/) are not versioned beyond `latest` and `stable`. It would be great to see releases have versioned docs on ReadTheDocs too.
* The examples in [`example-usecase/Yields_and_Resolutions.ipynb`](https://github.com/villano-lab/nrCascadeSim/blob/182e8a56406a8ae4e5cfc8df489d9779d25d93da/example-usecase/Yields_and_Resolutions.ipynb) are nice and it would be great to see these get added into the docs using some of the Jupyter rendering Sphinx extensions.

**Comment by lucydot** [comment]
Thank-you @matthewfeickert for your extremely thorough review! It is a 

@villaa it looks like it is over to you: the message above makes clear what actions are needed, https://github.com/villano-lab/nrCascadeSim/issues/75 is also a sticking point for @altavir. I can see that you and your team are currently working on this @villaa.

@altavir could I ask you to update your tick boxes at the top of the page, if you are in a position to do so? I can then better assess how close we are to completing the review process.

**Comment by matthewfeickert** [comment]
(I just went to open up Issues and I see that @gerudo7 was speedy and beat me to it. :slightly_smiling_face: Thanks! I'll add some comments on them to try to make them more helpful.)

**Comment by altavir** [comment]
@lucydot I've marked all but the State of the field box. I think that a comparison of the results is required for it.

**Comment by lucydot** [comment]
Hello @altavir - regarding state of the field: we do not usually require that there is a quantative comparison made with other  software. This section can include a qualitative comparison of functionality against other existing software (as is the case here), however if performance claims are made (e.g. that it runs faster) this must be backed up. As performance claims are not made for this code (from what I can see - please let me know if I am wrong), my view is that a comparison of results to other software is not needed.

However, as mentioned by Matthew above, and yourself in https://github.com/villano-lab/nrCascadeSim/issues/75, regression tests are needed to compare results from different versions of this software.

**Comment by altavir** [comment]
@lucydot the problem in this particular case is the only aim of this piece of software is a simulation of a specific process. Without a comparison, it is not possible to decide if the aim of the development is achieved. In simulations it is quite easy to get completely wrong result even with correct theory. Therefore, I believe that at least some comparison should be made. Eithere with an experiment data (it is better) or with other software. 

If the results are consistent with GEANT, but it worls much faster, it is OK. But if the results are significantly different, additional research is required to understand, which one is wrong (and GEANT could and is wrong on many occasions).

**Comment by villaa** [comment]
@lucydot @altavir we are committed to putting in the comparison with Geant4 at this point in any case. We also already have regression tests in the queue to be released. So, perhaps those things will be enough to satisfy all.

**Comment by lucydot** [comment]
@villaa I can see you are currently working on the Geant4 comparison testing (villano-lab/nrCascadeSim#75) - do you have a rough timescale for completion? Your software has travelled so far on the JOSS journey, I want to keep the train moving! 🚂

**Comment by gerudo7** [comment]
> @gerudo7 is working on finalizing the Geant4 comparison and The comparison plots have been finalized and from what I understand the work remaining is to get a reasonable exposition of these facts into the documentation and the paper, is this correct @gerudo7 ? I would expect this to take just a few days to finish, do you agree @gerudo7 ?

@villaa Yes. The Geant4 comparison feature branch is just about ready to be merged back, so I think a few days is a reasonable timescale to incorporate that information into the paper.

**Comment by villaa** [comment]
Ok @lucydot @altavir and @matthewfeickert we have just released `v1.4.0` -- the build system has been migrated to CMake and we have clearly documented comparisons to Geant4. These were the issues that were considered necessary for us to move on. 

Huge thanks to @gerudo7 who took care of the Geant4 comparison and generally did a lot of work to get our code up to par.

**Comment by altavir** [comment]
Looks good now. The comparison with GEANT4 shows a lot of differences, but it is for users to decide if the approximation is right or wrong in their case. The change of build configuration is also nice.

**Comment by matthewfeickert** [comment]
I'll try to get this done before Wednesday (2022-02-16) night as I had only minor suggestions but I will make sure this is reviewed before EOD on Friday (2022-02-18).

**Comment by matthewfeickert** [comment]
@lucydot Can you please have whedon set v1.4.0 as version, as that's the final one we're reviewing now?

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch v1.4.0. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03993/joss.03993/10.21105.joss.03993.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03993/joss.03993/10.21105.joss.03993.pdf) :page_facing_up: :point_left:

**Comment by matthewfeickert** [comment]
These comments are a follow up to my review on 2022-01-17 (https://github.com/openjournals/joss-reviews/issues/3993#issuecomment-1015075303) and pertain to release [v1.4.0](https://github.com/villano-lab/nrCascadeSim/releases/tag/v1.4.0) of the software.

## Notes

* [x] Change the `Makefile` to remove targeting `/usr/local` as a default. In many situations this is an unwritable location for users.

The change to CMake looks great and you have nice clean out-of-source builds now! :+1:

<details>
<summary>Example in Docker container:</summary>

```console
$ docker run --rm -ti -v $PWD:$PWD -w $PWD atlasamglab/stats-base:root6.24.06
root@ce876ebc60e7:/tmp/JOSS# git clone --depth 1 https://github.com/villano-lab/nrCascadeSim.git --branch v1.4.0 --single-branch
root@ce876ebc60e7:/tmp/JOSS# cmake -DCMAKE_INSTALL_PREFIX=/usr/local/venv -S nrCascadeSim -B build  # Using /usr/local/venv as it is already on PATH
root@ce876ebc60e7:/tmp/JOSS# cmake build -LH
root@ce876ebc60e7:/tmp/JOSS# cmake \
    --build build \
    --clean-first \
    --parallel $(($(nproc) - 1))
root@ce876ebc60e7:/tmp/JOSS# ls -lhtra build/
total 196K
drwxrwxr-x  4 1000 1000 4.0K Feb 19 03:50 ..
-rw-r--r--  1 root root  19K Feb 19 03:50 CMakeCache.txt
-rw-r--r--  1 root root 5.7K Feb 19 03:50 cmake_install.cmake
-rw-r--r--  1 root root  23K Feb 19 03:50 Makefile
-rwxr-xr-x  1 root root  25K Feb 19 03:50 regexPlayground
drwxr-xr-x  2 root root 4.0K Feb 19 03:50 lib
drwxr-xr-x  4 root root 4.0K Feb 19 03:50 .
-rwxr-xr-x  1 root root  99K Feb 19 03:50 realizeCascades
drwxr-xr-x 15 root root 4.0K Feb 19 03:50 CMakeFiles
root@ce876ebc60e7:/tmp/JOSS# cmake --build build --target install
root@ce876ebc60e7:/tmp/JOSS# command -v realizeCascades
/usr/local/venv/bin/realizeCascades
root@ce876ebc60e7:/tmp/JOSS# realizeCascades --help
Usage:  (null) options [ inputfile(s) ]
  -d, --seed          <integer>      Seed for random numbers
  -h, --help                         Print usage
  -n, --numgen        <number>       Number of traces to generate
  -o, --outfile       <filename>     Name the output file
  -s, --silent                       Silent, no standard out
  -v, --verbose       <level>        Print verbose messages at level <level>.
                                     Currently must use `--verbose=<level>` or `-v<level>` - no spaces.
  -V, --version                      Print version and exit
  -l, --log           <filename>     Log additional output to the specified file.
                                     If this option is not used, no logging will occur.
Report bugs to: villaa-at-gmail-dot-com
realizeCascades (nrCascadeSim) home page: <https://github.com/villano-lab/nrCascadeSim>
General help nrCascadeSim Software: <https://nrcascadesim.readthedocs.io/en/latest/>
root@ce876ebc60e7:/tmp/JOSS# time realizeCascades --verbose -n 100000 -o output.root nrCascadeSim/levelfiles/Si28_ngam_all_cascades_rfmt_sorted.txt
nrCascadeSim/levelfiles/Si28_ngam_all_cascades_rfmt_sorted.txt
Seed used: 3691720898
MT Result: 0.684558

real	0m1.881s
user	0m1.743s
sys	0m0.108s
root@ce876ebc60e7:/tmp/JOSS# cd build/
root@ce876ebc60e7:/tmp/JOSS/build# make uninstall
Scanning dependencies of target uninstall
Built target uninstall
root@ce876ebc60e7:/tmp/JOSS/build# ls /usr/local/venv/bin/realizeCascades
ls: cannot access '/usr/local/venv/bin/realizeCascades': No such file or directory
```

</details>

Small note: The [`v1.4.0` Download & Build docs](https://nrcascadesim.readthedocs.io/en/v1.4.0/01_Getting_Started.html#download-build) suggest to use the legacy in source build pattern and to change to Make right away

```
mkdir -p /path/to/build
cd /path/to/build
cmake -DCMAKE_INSTALL_PREFIX=/path/to/install /path/to/source
make
```

While not wrong, the legacy pattern isn't needed and even on old/stability focused operating systems like CentOS 7 in almost every situation in which software development is happening you'll have access to `devtoolset-7` or newer in which `cmake(3)` is available. So I'd recommend using the pattern of

```
$ cmake <options> -S <path to source> -B <path to build directory to make>
$ cmake <path to build directory> -LH  # See all of the options you've selected in a human readable form before build
$ cmake --build <path to build directory> --clean-first --parallel $(($(nproc) - 1))
```

as this gives the benefits of out-of-source builds, more transparency on what is happening, and allows for other build systems to be used other than Make (e.g. [Ninja](https://cmake.org/cmake/help/v3.23/generator/Ninja.html)).

For more on this please see the book [An Introduction to Modern CMake](https://cliutils.gitlab.io/modern-cmake/) by Henry Schreiner.

c.f. https://github.com/villano-lab/nrCascadeSim/issues/101

* [x] Add removal of the `.gcno` and `.gcda` files generated during the build to the `clean` Make target.
* [x] Remove recommendations of use of `sudo` to install software.

Note that the [`v1.4.0` Executables of `nrCascadeSim` docs](https://nrcascadesim.readthedocs.io/en/v1.4.0/02_Executables.html#using-realizecascades-command-line-executable) still have [one final instance of advocating for using `sudo make` that can be cleaned up](https://github.com/villano-lab/nrCascadeSim/blob/83fac469648dc940585c9eee64dbe9488e0fe434/docs/source/02_Executables.rst?plain=1#L35). Otherwise great!

c.f. https://github.com/villano-lab/nrCascadeSim/issues/102

* [x] Add at least one regression test to the test suite and plan to add more in the future.

* [x] In the State of the Field section it would be beneficial to mention that `Geant4` is open source software and to clarify if `FIFRELIN` is open source or not.

It would have been nice to clarify the `FIFRELIN` isn't an open source project, but I think the comparison that was done to `Geant4` `v10.7.3` is quite nice to have. :tada:

## Summary

@lucydot I am very impressed with the work that the team has done and I would give my full recommendation for publication with JOSS. I have opened two final GitHub issues as suggestions to the team for work in the future, but they should not be considered as further work for them to address in the review. :+1:

**Comment by lucydot** [comment]
@villaa congratulations 🎈 we are onto the very final steps now :) 

I'd like to give a huge thanks to @matthewfeickert and @altavir for their reviews. This software has been on a long and productive JOSS journey, steered forwards by your thoughtful reviews.

I'll do some checks now before passing this up to our editor-in-chief team for the final look over.

**Comment by lucydot** [comment]
@whedon set v1.4.0 as version

**Comment by whedon** [comment]
OK. v1.4.0 is the version.

**Comment by lucydot** [comment]
@whedon check references

**Comment by lucydot** [comment]
Excellent @villaa the paper looks good to me.

Please can you do the following:

- Make a tagged release of your software
- Archive this release on Zenodo (or similar)
        - The archive should have exactly the same title and author list as your JOSS paper
- Report the archive DOI and version number in this thread 👇

**Comment by villaa** [comment]
@lucydot Ok, we've done those things:

- tagged release: v1.4.0
- Zenodo DOI: 10.5281/zenodo.6056681
-

**Comment by lucydot** [comment]
Excellent!

**Comment by lucydot** [comment]
@whedon set 10.5281/zenodo.6056681 as archive

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1515/iupac.70.0026 is OK
- 10.1103/PhysRevD.91.083509 is OK
- 10.1103/physrevc.43.1086 is OK
- 10.1103/PhysRevC.46.972 is OK
- 10.1016/S0168-9002(03)01368-8 is OK
- 10.1088/1748-0221/16/07/p07032 is OK
- 10.1103/PhysRevC.82.054616 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by villaa** [comment]
@lucydot I don't know if the above is something I'm supposed to do or you. Also, that link seems to be the PR and I don't see where the final proof is in it. I did look at the proof from one of the previous commands and it looked good.

**Comment by matthewfeickert** [comment]
> I don't know if the above is something I'm supposed to do or you.

Nope you're good. One of the [journal editor-in-chiefs](https://joss.theoj.org/about#editorial_board) that is on duty this week will perform a final review of the draft generated on the joss-papers repo (linked above).

> Also, that link seems to be the PR and I don't see where the final proof is in it. I did look at the proof from one of the previous commands and it looked good.

Its there. It is just [a PDF artifact](https://github.com/openjournals/joss-papers/blob/ede3b57d6c84b6f942747b196a87bfa8c1129ea3/joss.03993/10.21105.joss.03993.pdf).

**Comment by arfon** [comment]
@editorialbot recommend-accept

**Comment by arfon** [comment]
@editorialbot recommend-accept

**Comment by arfon** [comment]
@editorialbot accept

**Comment by arfon** [comment]
@matthewfeickert, @altavir – many thanks for your reviews here and to @lucydot for editing this submission! JOSS relies upon the volunteer effort of people like you and we simply wouldn't be able to do this without you ✨ 

@villaa – your paper is now accepted and published in JOSS :zap::rocket::boom:

**Comment by matthewfeickert** [comment]
Congratulations and well done @villaa and @gerudo7! :tada:

**Comment by gerudo7** [comment]
Thank you! This is so exciting!

**Comment by villaa** [comment]
> Congratulations and well done @villaa and @gerudo7! 🎉

thanks @matthewfeickert we really appreciate your careful reviews and guidance. It was a bit long but the code is _much_ better for it!!
